src/c/c_client.cpp

```diff
@@ -30,517 +30,654 @@
 
 using namespace SmartRedis;
 
+
+// Return a pointer to a new Client.
+// The caller is responsible for deleting the client via DeleteClient().
 extern "C"
-void* SmartRedisCClient(bool cluster)
+void *SmartRedisCClient(bool cluster)
 {
-  /* Return a pointer to a new Client.
-  The user is responsible for deleting the client.
-  */
-  Client* s = new Client(cluster);
-  return (void*)s;
+  Client *s = NULL;
+
+  try {
+    s = new Client(cluster);
+  } catch (...) {
+    s = NULL;
+  }
+
+  return reinterpret_cast<void *>(s);
 }
 
+// Free the memory associated with the c client.
 extern "C"
-void DeleteCClient(void* c_client)
+void DeleteCClient(void *c_client)
 {
-  /* This function frees the memory associated
-  with the c client.
-  */
-  Client* s = (Client*)c_client;
+  // Sanity check params
+  if (NULL == c_client)
+    return; // Nothing we can do, so just bail. (error reporting to come later)
+
+  Client *s = reinterpret_cast<Client *>(c_client);
   delete s;
-  return;
 }
 
+// Put a dataset into the database.
 extern "C"
-void put_dataset(void* c_client, const void* dataset)
+void put_dataset(void *c_client, void *dataset)
 {
-  /* Put a dataset into the database.
-  */
-  Client* s = (Client *)c_client;
-  DataSet* d = (DataSet*)dataset;
+  // Sanity check params
+  if (NULL == c_client || NULL == dataset)
+    return; // Nothing we can do, so just bail. (error reporting to come later)
+
+  Client *s = reinterpret_cast<Client *>(c_client);
+  DataSet *d = reinterpret_cast<DataSet *>(dataset);
+
   s->put_dataset(*d);
-  return;
 }
 
+// Return a pointer to a new dataset.  The user is
+// responsible for deleting the dataset via DeallocateeDataSet()
 extern "C"
-void* get_dataset(void* c_client, const char* name,
-                  const size_t name_length)
+void *get_dataset(void *c_client, const char *name, const size_t name_length)
 {
-  /* Return a pointer to a new dataset.  The user is
-  responsible for deleting the dataset.
-  */
-  Client* s = (Client *)c_client;
-  std::string dataset_name = std::string(name, name_length);
-  DataSet* dataset = new DataSet(s->get_dataset(dataset_name));
-  return (void*)dataset;
+  // Sanity check params
+  if (NULL == c_client || NULL == name)
+    return NULL; // Nothing we can do, so just bail. (error reporting to come later)
+
+  Client *s = reinterpret_cast<Client *>(c_client);
+  std::string dataset_name(name, name_length);
+  DataSet *dataset = NULL;
+
+  try {
+    dataset = new DataSet(s->get_dataset(dataset_name));
```

Comment:
Right now the only exception ``new DataSet`` can throw is probably ``bad_alloc`` because we aren't doing anything except assigning string name.  However, to be safe and prevent memory leaks, maybe we split this one up with a ``bad_alloc`` catch and a ``...`` catch like the ``new client`` call?

Reply:
Agreed -- never know what can happen in the future